dvc/repo/gc.py

```diff
@@ -41,15 +42,27 @@ def gc(
 
         used = NamedCache()
         for repo in all_repos + [self]:
+            exclude_revs = (
+                repo.scm.list_branches()
+                if {"commits", "branches"} - set(remove)
+                else []
+            )
+            exclude_revs += (
+                repo.scm.list_tags()
+                if {"tags", "commits"} - set(remove)
+                else []
+            )
+
             used.update(
                 repo.used_cache(
-                    all_branches=all_branches,
+                    all_branches="branches" not in remove,
                     with_deps=with_deps,
-                    all_tags=all_tags,
-                    all_commits=all_commits,
+                    all_tags="tags" not in remove,
+                    all_commits="commits" not in remove,
```

Comment:
Need to not forget to remove `all_tags`, `all_branches` and `all_commits` from arguments for repo.gc